freebsd/utils/jail.cpp

```diff
@@ -0,0 +1,323 @@
+        name.erase(0, name.length() - limit);
+
+    return jail_name_prefix + name;
+}
+
+
+/// Create a jail with a given name and params string.
+///
+/// A new jail will always be 'persist', thus the caller is expected to remove
+/// the jail eventually via remove().
+///
+/// It's expected to be run in a subprocess.
+///
+/// \param jail_name Name of a new jail.
+/// \param jail_params String of jail parameters.
+void
+jail::create(const std::string& jail_name,
+             const std::string& jail_params)
+{
+    args_vector av;
+
+    // creation flag
+    av.push_back("-qc");
+
+    // jail name
+    av.push_back("name=" + jail_name);
+
+    // determine maximum allowed children.max
+    int max;
+    size_t len = sizeof(max);
+    if (::sysctlbyname("security.jail.children.max", &max, &len, NULL, 0) != 0) {
+        std::cerr << "sysctlbyname(security.jail.children.max) errors: "
+            << strerror(errno) << ".\n";
+        std::exit(EXIT_FAILURE);
+    }
+    if (len < sizeof(max)) {
+        std::cerr << "sysctlbyname(security.jail.children.max) provides less "
+            "data (" << len << ") than expected (" << sizeof(max) << ").\n";
+        std::exit(EXIT_FAILURE);
+    }
+    if (max < 0) {
+        std::cerr << "sysctlbyname(security.jail.children.max) yields "
+            "abnormal " << max << ".\n";
+        std::exit(EXIT_FAILURE);
+    }
+    if (max > 0)
+        max--; // a child jail must have less than parent's children.max
+    av.push_back("children.max=" + std::to_string(max));
+
+    // test defined jail params
+    const std::vector< std::string > params = parse_params_string(jail_params);
+    for (const std::string& p : params)
+        av.push_back(p);
+
+    // it must be persist
+    av.push_back("persist");
+
+    // invoke jail
+    std::auto_ptr< process::child > child = child::fork_capture(
+        run(fs::path("/usr/sbin/jail"), av));
+    process::status status = child->wait();
+
+    // expect success
+    if (status.exited() && status.exitstatus() == EXIT_SUCCESS)
+        return;
+
+    // otherwise, let us know what jail thinks and fail fast
+    std::cerr << child->output().rdbuf();
+    std::exit(EXIT_FAILURE);
+}
+
+
+/// Executes an external binary in a jail and replaces the current process.
+///
+/// \param jail_name Name of the jail to run within.
+/// \param program The test program binary absolute path.
+/// \param args The arguments to pass to the binary, without the program name.
+void
+jail::exec(const std::string& jail_name,
+           const fs::path& program,
+           const args_vector& args) throw()
+{
+    // get work dir prepared by kyua
+    char cwd[PATH_MAX];
+    if (::getcwd(cwd, sizeof(cwd)) == NULL) {
+        std::cerr << "jail::exec: getcwd() errors: "
+            << strerror(errno) << ".\n";
+        std::exit(EXIT_FAILURE);
+    }
+
+    // get jail id by its name
+    int jid = ::jail_getid(jail_name.c_str());
+    if (jid < 0) {
```

Comment:
"Explicit is better than implicit" (failing with `== -1` is documented in jail(3); other negative values are not).
```suggestion
    if (jid == -1) {
```
This also makes this call consistent with the other ones below.

Reply:
Good catch. Fixed.